### DISCUSSION TITLE
mpv-translate: Add translation to mpv via js scripts

### DISCUSSION THREAD

**Discussion by ahaoboy** [mpv-player/mpv#14976] [Show and tell]
![translate](https://github.com/user-attachments/assets/872ff037-465c-4a0d-963f-a5f6c854dd9b)

![interactive-translate](https://github.com/user-attachments/assets/42314e10-b0f0-4543-b4a3-cbade807c3d9)

## mpv-translate: A Translation Plugin for mpv-easy

mpv-translate is a translation plugin for mpv-easy. Due to its reliance on APIs not supported by the mpv JS script, it can only be used within mpv-easy. To get started, download the [mpy-easy-windows-full.zip](https://github.com/mpv-easy/mpv-easy/releases), which has only been tested on the Windows platform. You can activate the plugin using the shortcut key ```Ctrl + T```.

## How It Works

- Export Subtitles: Uses FFmpeg to extract subtitles from the current video and convert them into SRT format.

- Translation: Employs the Google Translate API to translate the SRT files into your desired language.

- Integration: Adds the translated subtitles back into your video.

Please note that the translation process can be slow, taking 5 to 10 seconds depending on network conditions. However, results are cached after the initial translation to improve efficiency.

## Future Improvements

Currently, mpv cannot directly obtain all subtitles, necessitating the use of FFmpeg, which increases the program size and does not support remote video. [How to extract subtitles to a file by using mpv](https://github.com/mpv-player/mpv/discussions/14942/)

Real-time translation experiences may suffer from network delays, and there is a risk of IP blocking.

The mpv JS script's lack of network API support means it cannot function as a standalone script.

## interactive-translate

Similar to [interSubs](https://github.com/oltodosel/interSubs), hovering over a word displays the word's info card, left mouse clicking on the word plays the word's audio, wheel clicking displays the line's translation.

You can use ```ctrl+i```to start interactive translation, which is still in its early stages, so for now, you can only use it in the mpv-easy project.
Currently only support English to Chinese, still need other language dictionary and translation api

github: https://github.com/mpv-easy/mpv-easy

**Comment by llyyr** [comment]
> Due to its reliance on APIs not supported by the mpv JS script

What is missing in mpv and what does "mpv-easy" add? If a script only works on a fork of mpv then I'm not sure the mpv discussions page is the right place to advertise it.

**Reply by ahaoboy** [reply]
mpv-easy is a toolkit for developing mpv scripts using ts. It is not a fork version of mpv. mpv uses mujs to run js scripts and does not add network module to it. Currently, it simulates the fetch function in js through the ```subprocess```

**Comment by guidocella** [comment]
> The mpv JS script's lack of network API support means it cannot function as a standalone script.

You can just call `curl` with `subprocess`, it works on every platform.

**Reply by ahaoboy** [reply]
Yes, currently I use subprocess to call local programs to simulate the fetch function in js, but I don’t use curl because Windows does not install this tool by default. I use rust to make a simple encapsulation.

**Reply by guidocella** [reply]
curl is installed by default since Windows 10.

**Reply by ahaoboy** [reply]
> curl is installed by default since Windows 10.

Thanks, this means we seem to have a cross-platform command line interface
